gallery_dl/extractor/iwara.py

```diff
@@ -0,0 +1,357 @@
+    """Extractor for iwara.tv search pages"""
+    subcategory = "search"
+    pattern = BASE_PATTERN + r"/search"
+    example = "https://www.iwara.tv/search?query=example&type=search_type"
+
+    def __init__(self, match):
+        IwaraExtractor.__init__(self, match)
+        parsed = urlparse(self.url)
+        parts = parsed.path.strip("/").split("/")
+        if len(parts) >= 1 and parts[0] == "search":
+            query_dict = parse_qs(parsed.query)
+            self.query = query_dict.get("query", [""])[0]
+            self.type = query_dict.get("type", [None])[0]
+        else:
+            return
+
+    def items(self):
+        collection = self.api.collection("/search", self.query)
+        if self.type == "video":
+            for video in collection:
+                video = self.api.item(f"/video/{video.get('id')}")
+                user_info = self.extract_user_info(video)
+                yield from self.yield_video(user_info, video)
+        elif self.type == "image":
+            for image_group in collection:
+                image_group = self.api.item(f"/image/{image_group.get('id')}")
+                if not image_group:
+                    return
+                user_info = self.extract_user_info(image_group)
+                yield from self.yield_image(user_info, image_group)
+
+
+class IwaraTagExtractor(IwaraExtractor):
+    """Extractor for iwara.tv tag search"""
+    subcategory = "tag"
+    pattern = BASE_PATTERN + r"/(videos|images)(?:\?.*)?"
+    example = "https://www.iwara.tv/videos?tags=example"
+
+    def __init__(self, match):
+        IwaraExtractor.__init__(self, match)
+        parsed = urlparse(self.url)
+        parts = parsed.path.strip("/").split("/")
+        if len(parts) >= 1 and parts[0] in ("videos", "images"):
+            query_dict = parse_qs(parsed.query)
+            self.type = parts[0]
+            self.tags = query_dict.get("tags", [""])[0]
+        else:
+            return
+
+    def items(self):
+        collection = self.api.collection(f"/{self.type}", self.tags)
+        if self.type == "videos":
+            for video in collection:
+                video = self.api.item(f"/video/{video.get('id')}")
+                user_info = self.extract_user_info(video)
+                yield from self.yield_video(user_info, video)
+        elif self.type == "images":
+            for image_group in collection:
+                image_group = self.api.item(f"/image/{image_group.get('id')}")
+                if not image_group:
+                    return
+                user_info = self.extract_user_info(image_group)
+                yield from self.yield_image(user_info, image_group)
+
+
+class IwaraAPI():
+    """Interface for the Iwara API"""
+    root = "https://api.iwara.tv"
+
+    def __init__(self, extractor):
+        self.extractor = extractor
+        self.token = None
+        self.headers = {}
+        self.login("/user/login")
+
+    def login(self, endpoint):
```

Comment:
Successful logins should be cached.

For an example, take a look at
https://github.com/mikf/gallery-dl/blob/d349c248c079cc70d043a8f312e427fe4c7c03a1/gallery_dl/extractor/bluesky.py#L523-L527